Jwt react frontend, suggested can be changed.

dgmouris opened the pull request.

dgmouris commented:
This Pull Requests refers to issue https://github.com/SimpleJWT/django-rest-framework-simplejwt/issues/263

Andrew-Chen-Wang commented:
Thanks for the PR! I'll take a look at the code, but I'll trust that the tests run smoothly and the UI looks good since I won't be installing React. I'll do my best to analyze the code and see if it meets requirements once I find the time within the next week.

dgmouris commented:
I didn't include a whole suite of tests here, just a heads up I will work to include them my apologies.

I'll include them in this pr, I'll work on this within the next few days.

Andrew-Chen-Wang commented on jwt-react/src/components/AuthenticatedPingTest/AuthenticatedPingTest.js:
```diff
@@ -0,0 +1,32 @@
+          <p>Ping Result: {pingResult} </p>
+        </React.Fragment>
+        :
+        <p>Please Login to see the ping test</p>
```
Only read on if the following is **IN**correct: if the user cannot hit the ping url, the Django server returns "Please login to see the ping test". Imo, I don't think that's true:

---

Hm "Please login to see the ping test"... if the user is logged out and the user is redirected to the log in page in which the `?next=` query parameter shows the page the user should go to on login, I don't think most people would like to see "Please login to see the ping test" (note "Login" should be lowercase).

In other words, this doesn't really make too much sense to me. If the user is forced to **RE**login, they shouldn't see "ping test" but rather the regular log in page.

Andrew-Chen-Wang commented on jwt-react/src/components/AuthenticatedPingTest/AuthenticatedPingTest.js:
```diff
@@ -0,0 +1,32 @@
+        <React.Fragment>
+          <h2>Authenticated Request Test</h2>
+          <button onClick={testPing}>Test Ping</button>
+          <p>Ping Result: {pingResult} </p>
```
`document.createTextNode()`?

dgmouris replied:
I think JSX is doing this under the hood. 

As per: https://reactjs.org/docs/introducing-jsx.html#jsx-prevents-injection-attacks

> By default, React DOM escapes any values embedded in JSX before rendering them. Thus it ensures that you can never inject anything that’s not explicitly written in your application. Everything is converted to a string before being rendered. This helps prevent XSS (cross-site-scripting) attacks.

I'll definitely take a deeper look at this as I think this might suffice, but definitely open to more discussion.

Andrew-Chen-Wang replied:
Oh wow that's nice to know. I think this comment be be resolved.

Andrew-Chen-Wang commented on jwt-react/src/api/auth.js:
```diff
@@ -0,0 +1,102 @@
+      originalRequest.headers['Authorization'] = headerAuthorization;
+      return authRequest(originalRequest)
+    }).catch((error)=> {
+      // Note If you are using browser router you can redirect here
```
Slightly confused when you mean browser router; is this some kind of React library? If so, when you mean redirect here, do you mean "re-login" here? If so, this would be a nice place to rewrite `window.location.href = domain + "login/?next=CURRENT_PATH"` or whatever. Basically show a login page.

dgmouris replied:
Again this another instance of adding `react-router-dom` to the project. Will add with some pages to make this more thorough.

Andrew-Chen-Wang reviewed (changes requested):
Thanks the the PR and the language PR @dgmouris. ~~Once I get my hw done, hopefully by the weekend, I can review this awesome PR. I took a quick glance at it, and it seems to be doing what I asked for, but just to double check:~~ Scratch that, here's my review (quick note, for my 4 suggestions, go to the files tab and press "Add suggestion to batch" rather than committing them one at a time):

 - [X] I believe this is done properly: On 401, refresh the access token (i.e. get new access token)
 - [ ] If the above didn't work, automatically log out the user and show a login page.
 - [x] If the user is logged out, append `?next=current_path` so that the next link the user goes to is the original link the user went before being forced to go to the login url.

Andrew-Chen-Wang commented:
Just took a look at package.json (I knew what was going to be in there, but had to be sure :P). Out of curiosity, is there a reason why some integrities still use the NIST deprecated SHA1? Maybe it's because the packages haven't been updated since Google broke it?

dgmouris commented on server/.gitignore:
```diff
@@ -1,124 +1,27 @@
-# Byte-compiled / optimized / DLL files
```
My apologies on this, I'm still getting a feel for what you folks would like on the is project as it's fairly new. 

I'll just add the contents from the previous commits very soon. Thanks for letting me know:)

dgmouris commented on server/server/settings.py:
```diff
@@ -162,3 +163,10 @@
+CORS_ALLOW_CREDENTIALS = True
+
+# Below needs to be changed for production
+CORS_ORIGIN_WHITELIST = ['http://localhost:3000']
```
I'll add an example, that's a great suggestion. 

Basically this would be wherever your deploy your react application, I'll definitely add a sample production url (which I think is what you're looking for on this). I might be misunderstanding what you're asking for here, so please let me know if that's the case.

Andrew-Chen-Wang replied:
I believe that's exactly right, just adding a production level domain. Since there's no separation between settings files for different environments, I guess it's fine to leave a comment rather than adding it to the list.

dgmouris commented on jwt-react/src/components/App/App.js:
```diff
@@ -0,0 +1,26 @@
+
+  return (
+    <div className="App">
+      <UserContext.Provider value={{user, setUser, isUserLoggedIn}}>
```
Although I agree with escaping this later in this pr, on I don't think that "user" should be escaped in this instance it's purely just passing "user" as an argument so that it can be used with the "useContext" hook of react (basically so that user can be used in child components). 

Hopefully that makes sense, I'm definitely open to change, but the above is something to consider. :smile:

Andrew-Chen-Wang replied:
```suggestion
      // since the values are stored as variables in context, it's not necessary to escape the values for XSS-sake.
      <UserContext.Provider value={{user, setUser, isUserLoggedIn}}>
```

I see. That makes sense. I'm not totally aware of what `Context` means in React land, but I think I got the gist of it. Thanks for comparing this to the other part of the PR that I reviewed. I think it's reasonable to perhaps leave a suggestion as a comment as to why escaping is not necessary?

Andrew-Chen-Wang replied:
```suggestion
      // since the values are stored as variables in context, it's not necessary to escape the values for XSS-sake.
      <UserContext.Provider value={{user, setUser, isUserLoggedIn}}>
```

I see. That makes sense. I'm not totally aware of what `Context` means in React land, but I think I got the gist of it. Thanks for comparing this to the other part of the PR that I reviewed. I think it's reasonable to perhaps leave a suggestion as a comment as to why escaping is not necessary?

Andrew-Chen-Wang replied:
Per your other comment regarding JSX, would user, in this case, be automatically escaped?

Andrew-Chen-Wang reviewed (changes requested):
I think it might be useful in the README to state that, "unlike with regular Django templating, make sure you HTML escape anything being placed in the DOM to prevent XSS."

dgmouris commented on jwt-react/src/components/Login/Login.js:
```diff
@@ -0,0 +1,72 @@
+    {
+      isUserLoggedIn() ?
+      <React.Fragment>
+      <h3>{user.username} is currently Logged in</h3>
```
Hey thanks for the review on this, I wasn't sure really the scope of this. 

I didn't add the `react-router-dom` (https://www.npmjs.com/package/react-router-dom) because I thought it would fall out of scope. Looking as this idea was referenced a couple of times I will modify this accordingly and add a couple of pages.

dgmouris commented:
> Just took a look at package.json (I knew what was going to be in there, but had to be sure :P). Out of curiosity, is there a reason why some integrities still use the NIST deprecated SHA1? Maybe it's because the packages haven't been updated since Google broke it?

I'll definitely have to take a look at this I'm not sure why they're using this.

dgmouris commented on jwt-react/src/api/auth.test.js:
```diff
@@ -0,0 +1,94 @@
+      .mockReturnValueOnce(REFRESH_TOKEN_VALUE) // for refreshToken
+      .mockReturnValueOnce(ACCESS_TOKEN_VALUE) // for the error interceptor
+
+    return errorInterceptor({config: errorConfig, response: ERROR_RESPONSE})
+      .finally(()=> {
+        expect(tokenRequest.post).toHaveBeenCalledWith('/api/token/access/',
+          {"refresh": REFRESH_TOKEN}
+        );
+        expect(errorConfig.headers['Authorization']).toEqual(`Bearer ${ACCESS_TOKEN_VALUE}`)
+        expect(authRequest).toHaveBeenCalledWith(errorConfig)
+      });
+  });
+});
```
Will add this test:)

Andrew-Chen-Wang replied:
Ping @dgmouris Was this implemented? Couldn't find a commit for it.

dgmouris commented on jwt-react/src/api/rest.js:
```diff
@@ -0,0 +1,17 @@
+import { authRequest } from './auth.js'
+
+const ping = () => {
+  const extraParameters = {params: {"id": "PONG" }}
+  return authRequest.get('/api/ping/',extraParameters)
+    .then(response=>{
+      if (response === undefined) {
+        throw new Error("User has not logged in yet")
+      }
+      return Promise.resolve(response.data)
+    }).catch((error)=> {
+      return Promise.reject(error)
+    });
+}
```
I just removed the lines as they straight up aren't necesarry.

dgmouris commented:
Hey @Andrew-Chen-Wang, 

I've added a bunch of changes that can be summarized as such based on suggested changes:

- Added `react-router-dom` to provide "page" functionality and implemented the "login?next="/ping" to route to correct page after login.
- Added react bootstrap to make it an acceptable looking example when people run it locally.
- Added the domain on the CORS_WHITELIST
- Comments and other smaller changes suggested

Things I haven't changed:
- Since this a demo that I've used create react app the service worker just comes out of the box.
- Didn't change the bit.ly links as those just came out of the box with create-react-app.

I hope this is now acceptable to merge I think that it should be all good. Let me know if you have more thoughts:)

This PR is getting a bit massive, so I think the XSS cookie functionality that you mentioned should be a todo later, I can open an issue if you would like.

Thanks again!

Andrew-Chen-Wang commented on jwt-react/src/serviceWorker.js:
```diff
@@ -0,0 +1,141 @@
+    window.addEventListener('load', () => {
+      const swUrl = `${process.env.PUBLIC_URL}/service-worker.js`;
+
+      if (isLocalhost) {
```
Ping @dgmouris Anything on this?

dgmouris replied:
Hey @Andrew-Chen-Wang,

Hopefully this puts your mind at ease

> Service workers require HTTPS, although to facilitate local testing, that policy does not apply to localhost. If your production web server does not support HTTPS, then the service worker registration will fail, but the rest of your web app will remain functional.

Source: https://create-react-app.dev/docs/making-a-progressive-web-app/#offline-first-considerations

I think this answers your question.

Andrew-Chen-Wang replied:
Awesome, thank you!

Andrew-Chen-Wang commented:
Hey @dgmouris Not sure if you got the pings, but I've resolved all comments except the ones that you didn't address or should address. Additionally, I still can't tell if the user is automatically logged out after the refresh token is expired (can't find it code wise or test wise. Let me know if I'm wrong).

Thanks for pouring so much time and effort into this! Once those comments are resolved, it's good to be merged!

dgmouris commented:
Hey @Andrew-Chen-Wang, 

My apologies on that. It should be fixed with tests in the commit https://github.com/SimpleJWT/drf-SimpleJWT-React/pull/1/commits/689bc813f09f64aaf11aab5af30129832c59bdd7 if you'd like take a gander:)

I'll take a look at your other comments and hopefully resolve them shortly.

Andrew-Chen-Wang commented:
Awesome work @dgmouris, thanks for implementing this!
